asim: output test set up if eval has full = true

wenyihu6 opened the pull request:
Resolves: https://github.com/cockroachdb/cockroach/issues/153572
Release note: none

---
**asim: add a separator line after each run**

This commit updates asim output to add a separate line after each simulation run
to improve readability.

---
**asim: export DefaultStartTime**

This commit exports the simulator’s default start time, which will be used in
future commits to compare against scheduled event timestamps in the scheduler
package, ensuring timestamp is printed only when not at the start time.

---
**asim: move some variables to an outer scope**

This commit moves certain variables above the configuration declarations so they
can be reused in an outer scope.

---
**asim: generate {TESTNAME}_setup.txt**

This commit adds a test setup string which will be outputted to
{TESTNAME}_setup.txt. Only the first configuration and first sample are emitted,
since the setup should effectively be the same across runs (other than
LBRebalancingMode) and duplicating it would add unnecessary runtime churn. One
downside is that the scheduled event for LBRebalancingMode will only reflect the
first occurrence. If a datadriven test contains multiple evals, the last one
wins.

---
**asim: remove run.stateStrAcrossSamples append**

This commit removes an unused run.stateStrAcrossSamples append. I left the field
itself in place in the struct since it causes no harm, though we can consider
deleting it entirely.

---
**asim: print simulation setting as part of * _setup.txt**

This commit includes simulation settings in the setup output. Only values that
differ from the default simulation settings are printed. Note that cluster
settings and the seed are omitted. The simulation seed is derived from
rand.New(rand.NewSource(42)).Int63() by default, while the default simulation
setting itself uses 42.

Note that we rely on users to explicitly reset cluster settings between runs. It
isn’t an issue now, but it may become a potential footgun in the future.

---
**asim: print store info to * _setup.txt**

This commit outputs store capacity and store attributes t.o the setup file.

---
**asim: output test set up if eval has full = true**

This commit adds an optional parameter to eval that outputs the test setup. By
default, the setup is only written to the generated folder as
{TESTNAME}_setup.txt.

wenyihu6 committed:
asim: generate {TESTNAME}_setup.txt

This commit adds a test setup string which will be outputted to
{TESTNAME}_setup.txt. Only the first configuration and first sample are emitted,
since the setup should effectively be the same across runs (other than
LBRebalancingMode) and duplicating it would add unnecessary runtime churn. One
downside is that the scheduled event for LBRebalancingMode will only reflect the
first occurrence. If a datadriven test contains multiple evals, the last one
wins.

wenyihu6 committed:
asim: add a separator line after each run

This commit updates asim output to add a separate line after each simulation run
to improve readability.

wenyihu6 committed:
asim: export DefaultStartTime

This commit exports the simulator’s default start time, which will be used in
future commits to compare against scheduled event timestamps in the scheduler
package, ensuring timestamp is printed only when not at the start time.

wenyihu6 committed:
asim: move some variables to an outer scope

This commit moves certain variables above the configuration declarations so they
can be reused in an outer scope.

wenyihu6 committed:
asim: remove run.stateStrAcrossSamples append

This commit removes an unused run.stateStrAcrossSamples append. I left the field
itself in place in the struct since it causes no harm, though we can consider
deleting it entirely.

wenyihu6 committed:
asim: print simulation setting as part of * _setup.txt

This commit includes simulation settings in the setup output. Only values that
differ from the default simulation settings are printed. Note that cluster
settings and the seed are omitted. The simulation seed is derived from
rand.New(rand.NewSource(42)).Int63() by default, while the default simulation
setting itself uses 42.

Note that we rely on users to explicitly reset cluster settings between runs. It
isn’t an issue now, but it may become a potential footgun in the future.

wenyihu6 committed:
asim: print store info to * _setup.txt

This commit outputs store capacity and store attributes t.o the setup file.

wenyihu6 committed:
asim: output test set up if eval has full = true

This commit adds an optional parameter to eval that outputs the test setup. By
default, the setup is only written to the generated folder as
{TESTNAME}_setup.txt.

wenyihu6 commented:
Some of the formatting may have one-off errors and not very principled. I’ll leave it as is for now since it’s only for formatting. We can revisit it later if it becomes an issue.

tbg reviewed (changes requested):
> @tbg reviewed 29 of 29 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 22 of 22 files at r4, 1 of 3 files at r5, 8 of 8 files at r6, 5 of 5 files at r7, 4 of 4 files at r8, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153545)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt` line 96 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_GuqBV7_tt44ZmLLhL:-O_GuqBV7_tt44ZmLLhM:bsb0wtl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0660dcda4b042fad8791713228fdeddfe97bc4c5/pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt#L96)):*
> ```help
> 	[0,200000): 10(rf=3), 0MiB
> 	[0,200000): write-only [1B/op, 0ops/s]
> 	
> ```

please make sure there aren't any extra newlines and datadriven doesn't fall into the newline-escaping syntax.
___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt` line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_Gu83IDOv9DCsngU6p:-O_Gu83IDOv9DCsngU6q:b5bmgxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3195dc974a39c5d48d82b1b261e4327db29b4c60/pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt#L78)):*
> ```help
> 	[nodes: 3, stores_per_node:1, store_disk_capacity: 256GiB, node_capacity: 0cpu-sec/sec]
> 	[0,200000): 10(rf=3), 0MiB
> 	[0,200000): write-only [1B/op, 0ops/s]	
> ```

it might be hard for the uninitiated to surmise that this is a write load while the similar-looking line above is the range definition. No need to change anything now, just musing.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
asim: fix one-off formatting in rand testing

Previously, some one-off formatting introduced extra lines in the randomized
testing output. This commit fixes the issue and makes the formatting approach
more consistent and principled.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153545)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt` line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_Gu83IDOv9DCsngU6p:-O_J4v0kEDNXnVCVKqV_:b55jen1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3195dc974a39c5d48d82b1b261e4327db29b4c60/pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt#L78)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

it might be hard for the uninitiated to surmise that this is a write load while the similar-looking line above is the range definition. No need to change anything now, just musing.
</blockquote></details>

Ack, we can improve as we go in the future. 
___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt` line 96 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_GuqBV7_tt44ZmLLhL:-O_J4y3n1aKFJTnjI4yf:b-5zotnj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0660dcda4b042fad8791713228fdeddfe97bc4c5/pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt#L96)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

please make sure there aren't any extra newlines and datadriven doesn't fall into the newline-escaping syntax.
</blockquote></details>

:see_no_evil: Should be better now.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
Looks like there's a merge conflict now, but otherwise LGTM!

> @tbg reviewed 12 of 12 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-:-O_Li74679wkFTDrb1Is:bm0meif)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt` line 96 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_GuqBV7_tt44ZmLLhL:-O_LhwuJ9IYVH4e0r2kl:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0660dcda4b042fad8791713228fdeddfe97bc4c5/pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt#L96)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

:see_no_evil: Should be better now.
</blockquote></details>

Thanks!
___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt` line 79 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_LhnuGCoG2agYecraq:-O_LhnuGCoG2agYecrar:b1a11d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/040760373ed9dbbc1c77f59e40ac7c35a0e35c79/pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt#L79)):*
> ```help
> 	[0,200000): 10(rf=3), 0MiB
> 	[0,200000): write-only [1B/op, 0ops/s]
> 	empty
> ```

Like this is now going to be pretty confusing :laughing: Likely each of these lines should have a prefix for what it is that it's printing. (Not for this PR, but maybe you could follow up?)
___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt` line 119 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_LhzYFAZb4dj40AWNd:-O_LhzYFAZb4dj40AWNe:b-k2lgsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/040760373ed9dbbc1c77f59e40ac7c35a0e35c79/pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt#L119)):*
> ```help
> sample3: start running
> initial state at 2022-03-21 11:00:00:
> 	stores(3)=[s1n1=(replicas(10)),s2n2=(replicas(10)),s3n3=(replicas(10))]
> ```

This is intentional, right?


<!-- Sent from Reviewable.io -->

tbg commented:
PR title could use an update.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153545)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt` line 79 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_LhnuGCoG2agYecraq:-O_Mrm4X8JZmC4Os_TGP:b-3eenhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/040760373ed9dbbc1c77f59e40ac7c35a0e35c79/pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt#L79)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Like this is now going to be pretty confusing :laughing: Likely each of these lines should have a prefix for what it is that it's printing. (Not for this PR, but maybe you could follow up?)
</blockquote></details>

Sg, I will follow up. I’m using `\t` for some of the strings without explicitly passing a tag from the caller, so rand testing is still taking the loss from the one-off sometimes. I will try to clean it up in the follow up. 
___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt` line 119 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_LhzYFAZb4dj40AWNd:-O_Msq2i17yfPYlkOxtZ:b-7qe1nm) ([raw file](https://github.com/cockroachdb/cockroach/blob/040760373ed9dbbc1c77f59e40ac7c35a0e35c79/pkg/kv/kvserver/asim/tests/testdata/rand/default_settings.txt#L119)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is intentional, right?
</blockquote></details>

Not really intentional, but I was fine with settling on this.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Rebased on master and resolved the conflicts.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153545)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/asim/tests/testdata/rand/weighted_rand.txt` line 73 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/153545#-O_Pj4aM1i10VUh64gLc:-O_Pj4aN48sdW-LLpR2i:b-t7qvlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7fe80db7a9db59cb3692c206ee9a4ccd9df8192/pkg/kv/kvserver/asim/tests/testdata/rand/weighted_rand.txt#L73)):*
> ```help
> 	[nodes: 3, stores_per_node:2, store_disk_capacity: 256GiB, node_capacity: 0cpu-sec/sec]
> 	weighted randomized ranges with placement_type=weighted_rand, weighted_rand=[0 0 0 0.3 0.3 0.4], {563 0 160411 3 0 }
> 	[0,200000): write-only [1B/op, 0ops/s]
> ```

Hm, this is certainly unintentional. I will address this in the follow up as well :face_with_open_eyes_and_hand_over_mouth: 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r=tbg
